fuel-crypto/src/message.rs

```diff
@@ -12,54 +10,39 @@ use core::ops::Deref;
 pub struct Message(Bytes32);
 
 impl Message {
-    /// Memory length of the type
+    /// Memory length of the type in bytes.
     pub const LEN: usize = Bytes32::LEN;
 
-    /// Normalize a message for signature
+    /// Normalize the given message by cryptographically hashing its content in
+    /// preparation for signing.
     pub fn new<M>(message: M) -> Self
     where
         M: AsRef<[u8]>,
     {
         Self(Hasher::hash(message))
     }
 
-    /// Add a conversion from arbitrary slices into owned
+    /// Construct a `Message` directly from its bytes.
     ///
-    /// # Safety
-    ///
-    /// There is no guarantee the provided bytes will be the product of a cryptographically secure
-    /// hash. Using insecure messages might compromise the security of the signature.
-    pub unsafe fn from_bytes_unchecked(bytes: [u8; Self::LEN]) -> Self {
+    /// This constructor expects the given bytes to be a valid,
+    /// cryptographically hashed message. No hashing is performed.
+    pub fn from_bytes(bytes: [u8; Self::LEN]) -> Self {
         Self(bytes.into())
     }
 
-    /// Add a conversion from arbitrary slices into owned
-    ///
-    /// # Safety
-    ///
-    /// This function will not panic if the length of the slice is smaller than
-    /// `Self::LEN`. Instead, it will cause undefined behavior and read random
-    /// disowned bytes.
+    /// Construct a `Message` reference directly from a reference to its bytes.
     ///
-    /// This function extends the unsafety of [`Self::from_bytes_unchecked`].
-    pub unsafe fn from_slice_unchecked(bytes: &[u8]) -> Self {
-        Self(Bytes32::from_slice_unchecked(bytes))
+    /// This constructor expects the given bytes to be a valid,
+    /// cryptographically hashed message. No hashing is performed.
+    pub fn from_bytes_ref(bytes: &[u8; Self::LEN]) -> &Self {
```

Comment:
Does this actually uphold the lifetime. If the bytes array is moved after this call and `&Self` is still being used, do we get a compile error? If not this should still be an unsafe function right?

Reply:
> Does this actually uphold the lifetime.

Yes, in this case the lifetime of the reference in `&Self` is bound by the only input lifetime, i.e. the lifetime of `bytes: &[u8; Self::LEN]`. In this case, Rust elides the lifetime because there is only one possible valid lifetime for the output. Here is what the un-ellided signature would look like:

```rust
    pub fn from_bytes_ref<'a>(bytes: &'a [u8; Self::LEN]) -> &'a Self {
```